api/raftlease/client.go

```diff
@@ -0,0 +1,609 @@
+				// If there are no addresses, then nothing is routable. In this
+				// case, we'll continue to use the current addresses.
+				c.config.Logger.Errorf("no server addresses found, will continue to use old addresses")
+				continue
+			}
+
+			if err := c.ensureServers(addresses); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (c *Client) initServers() error {
+	if len(c.config.APIInfo.Addrs) == 0 {
+		return errors.NotFoundf("api addresses")
+	}
+
+	for k, address := range c.config.APIInfo.Addrs {
+		info := *c.config.APIInfo
+		info.Addrs = []string{address}
+
+		remote := c.config.NewRemote(RemoteConfig{
+			APIInfo: &info,
+			Clock:   c.config.Clock,
+			Logger:  c.config.Logger,
+		})
+
+		// In reality it doesn't matter what these get called, as a later down
+		// the line they'll be replaced via the request from the
+		// apiserver.Details result.
+		key := fmt.Sprintf("%d", k)
+		c.servers[key] = remote
+	}
+
+	return nil
+}
+
+// gatherAddresses turns a series of data addresses into a map of server ids
+// and internal address. If no internal address is found, then fallback to
+// the external sorted addresses.
+func (c *Client) gatherAddresses(details apiserver.Details) map[string]string {
+	if len(details.Servers) == 0 {
+		return nil
+	}
+
+	servers := make(map[string]string)
+	for id, server := range details.Servers {
+		var address string
+		if server.InternalAddress != "" {
+			address = server.InternalAddress
+		} else if len(server.Addresses) > 0 {
+			// The sorting of the addresses is done during the publishing of
+			// the event, so we can depend on the correct ordering.
+			address = server.Addresses[0]
+		}
+
+		// We can't use this server, so skip it.
+		if address == "" {
+			continue
+		}
+		servers[id] = address
+	}
+	return servers
+}
+
+func (c *Client) ensureServers(addresses map[string]string) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	witnessed := set.NewStrings()
+	for id, address := range addresses {
+		witnessed.Add(id)
+
+		// If we already have a server, don't tear it down, just update the
+		// address.
+		if server, found := c.servers[id]; found {
+			server.SetAddress(address)
+		} else {
+			info := *c.config.APIInfo
+			info.Addrs = []string{address}
+
+			remote := c.config.NewRemote(RemoteConfig{
+				APIInfo: &info,
+				Clock:   c.config.Clock,
+				Logger:  c.config.Logger,
+			})
+			c.servers[id] = remote
+			if err := c.catacomb.Add(remote); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	// Kill off any servers that have fallen away. In a highly available setup,
+	// this might actually be really painful for us. We should really consider
+	// if this is wise or not? During flapping instances, we take a long time
+	// to connect and this could cause stalling of message sending.
+	//
+	// Instead we should probably just tombstone them first time around, then
+	// next time around we should reap them.
+	for id, remote := range c.servers {
```

Comment:
To handle flapping instances you could add these into a reap set (for this case a map[server]goneTimestamp should suffice) which gets periodically reaped in your main loop (i.e. kill anything where goneTimestamp < threshold). In the meantime, if you happen to see the server again just remove it from the reap set.